test/functional/p2p_invalid_messages.py

```diff
@@ -42,13 +43,30 @@ def set_test_params(self):
         self.setup_clean_chain = True
 
     def run_test(self):
+        self.test_buffer()
         self.test_magic_bytes()
         self.test_checksum()
         self.test_size()
         self.test_msgtype()
         self.test_large_inv()
         self.test_resource_exhaustion()
 
+    def test_buffer(self):
+        self.log.info("Test message with header split across two buffers")
+        conn = self.nodes[0].add_p2p_connection(P2PDataStore())
+        # Create valid message
+        msg = conn.build_message(msg_ping(nonce=12345))
+        cut_pos = 12    # Chosen at an arbitrary position within the header
+        # Send message in two pieces
+        before = self.nodes[0].getnettotals()['totalbytesrecv']
+        conn.send_raw_message(msg[:cut_pos])
+        middle = self.nodes[0].getnettotals()['totalbytesrecv']
+        # If this assert fails, we've hit an unlikely race
+        # where the test framework sent a message in between the two halves
+        assert middle == before + cut_pos
```

Comment:
use `assert_equal`

Reply:
Why not test the debug log output as well. I think this works; maybe it can be improved on:

```diff
-        msg = conn.build_message(msg_ping(nonce=12345))
-        cut_pos = 12    # Chosen at an arbitrary position within the header
-        # Send message in two pieces
-        before = self.nodes[0].getnettotals()['totalbytesrecv']
-        conn.send_raw_message(msg[:cut_pos])
-        middle = self.nodes[0].getnettotals()['totalbytesrecv']
-        # If this assert fails, we've hit an unlikely race
-        # where the test framework sent a message in between the two halves
-        assert middle == before + cut_pos
-        conn.send_raw_message(msg[cut_pos:])
-        conn.sync_with_ping(timeout=1)
+        with self.nodes[0].assert_debug_log(['[net] Incomplete headerReceived']):
+            msg = conn.build_message(msg_ping(nonce=12345))
+            cut_pos = 12    # Chosen at an arbitrary position within the header
+            # Send message in two pieces
+            before = self.nodes[0].getnettotals()['totalbytesrecv']
+            conn.send_raw_message(msg[:cut_pos])
+            middle = self.nodes[0].getnettotals()['totalbytesrecv']
+            # If this assert fails, we've hit an unlikely race
+            # where the test framework sent a message in between the two halves
+            assert middle == before + cut_pos
+            conn.send_raw_message(msg[cut_pos:])
+            conn.sync_with_ping(timeout=1)

with self.nodes[0].assert_debug_log(['[net] Incomplete headerReceived']):
```

Reply:
See below

Reply:
It tests the logging, which is a user-facing functionality.

Reply:
Oh right, if it's just sockets then agreed.